Material write hdf5 refactor

bam241 opened the pull request:
this updates the write_hdf5 method to refactor the structure of material in an hdf5 file as such:

```
/material/
/material/mat_1/
/material/mat_1/nucid
/material/mat_1/composition
/material/mat_1/composition_metadata
```

`/material` and `/datapath` being HDF5 groups
 `/nucid`, `composition` and `composition_metadata` being HDF5 dataset

the value of `datapath` is string variable can/should be user provided.

this will allows a future MaterialLibrary (see #1104) to coexist with simple material inside the same hdf5 file

this is not done yet, still need some cleaning ad missing docstring.
I also need to tweak the `from_hdf5` method to be able to read the new structure


but for some reason `nosetests` are not finishing on my computer for this (or `master`....)

bam241 committed:
add wrtie_nuc_methods

bam241 committed:
starts to add new write_hdf5

bam241 committed:
add std writing methods

bam241 committed:
add some dostring

bam241 commented:
@gonuke is it something like that you had in mind ?

gonuke commented:
Where will material libraries live in this structure?

bam241 commented:
material library will live in:
`/material_library`
so a multiple single materials and multiple material_library could co-exist in the same file


so something like:
```
/material_library/
/material_library/mylib_1/
/material_library/mylib_1/nucid
/material_library/mylib_1/composition
/material_library/mylib_1/composition_metadata
/material_library/mylib_2/
/material_library/mylib_2/nucid
/material_library/mylib_2/composition
/material_library/mylib_2/composition_metadata
```

with `datapath` being a `string` variable

bam241 committed:
fixing bad loop

bam241 committed:
clean

bam241 committed:
adding possibility to specify nucpath in _protocol1_read

bam241 committed:
shoudl be working

bam241 committed:
add some test, fix the existing ones

bam241 committed:
simplify write_hdf5

bam241 committed:
adding docstring

bam241 committed:
adding news file

bam241 commented:
@gonuke @pshriwise @makeclean, I think this is ready for review!

I issue this there and not in the material_library PR (#1104 ) to simplify the review process.

Similar structure refactoring will be done for the `MaterialLibrary`


a simple test has been added to ensure, backward compatibility and ability to still use the old structure

bam241 committed:
add missing word

bam241 committed:
fix inent

bam241 commented:
@ngranda @ypark234 @a-co @ NybergWISC : feel free to review this if you have 5 minutes to spear :)

This is mainly some C++ development, with some hint of `python` binding using `Cython`

a-co reviewed (commented):
The loop looks good to me. Nice solution to avoid the default false.

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +480,125 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+        // create Group
+        H5Gcreate2(db, "/material", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+      } else if (object_info.type != H5O_TYPE_GROUP) {
+        std::cout << "Non-group/non-dataset object /Material already exists "
```
```suggestion
        std::cout << "Non-group/non-dataset object /material already exists "
```

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +480,125 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+        // no Group but path fallback on old method using default
+        if (datapath_exists) {
+          H5Fclose(db);
+          write_hdf5(filename, datapath, "/nucid", row, chunksize);
```
is it always "/nucid" in this case?

gonuke replied:
Could there be a pre-existing nucpath that has a different name?  Maybe it doesn't matter because now it will definitely be this name?

bam241 replied:
this is true, We should attempt to discover it....

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +480,125 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+    if (object_info.type == H5O_TYPE_DATASET)
+       {
+        // Check if path exists
+        bool datapath_exists = h5wrap::path_exists(db, datapath);
```
why not just make the call in the `if` statement below?

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +480,125 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+        // Check if path exists
+        bool datapath_exists = h5wrap::path_exists(db, datapath);
+        // no Group but path fallback on old method using default
+        if (datapath_exists) {
```
what do we do if this is a dataset, but the datapath doesn't exist?

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +480,125 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+          write_hdf5(filename, datapath, "/nucid", row, chunksize);
+          return;
+        }
+      } else if ( object_info.type >= H5O_TYPE_NTYPES || object_info.type < H5O_TYPE_UNKNOWN ) {
```
are these the conditions that define that a group does not already exist?  Is there no simpler way?

gonuke replied:
This may need more comments for which case is which in the if statements

gonuke reviewed (changes requested):
This all looks a little messy to me.  I think the logic can be cleaned up and/or documented better to make it easier to follow.  I understand that some of this comes from wanting to preserve the backwards compatibility, but maybe we should do that in a more modular way to better communicate to future developers.

bam241 committed:
Apply suggestions from code review

Co-Authored-By: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
simplify write_hdf5 + use relative path instead of forming the absolute ones

bam241 committed:
fix bug readd absolute path :(

bam241 commented on pyne/cpp_material.pxd:
```diff
@@ -53,9 +53,8 @@ cdef extern from "material.h" namespace "pyne":
-        void write_hdf5(char *, char *, char *, float) except +
-        void write_hdf5(char *, char *, char *, float, int) except +
+        void deprecated_write_hdf5(char *, char *, char *, float, int) except +
+        void write_hdf5(char *, char *, float, int) except +
```
Here I decided to expose only the C++ method we are actually using on the python binding (for `write_hdf5()`)

bam241 commented on src/material.cpp:
```diff
@@ -469,13 +497,149 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+  //std::string nucpath = "/nuclidelist";
+  std::string full_datapath = "/material" + datapath + "/composition";
+  std::string nucpath = "/material" + datapath + "/nuclidelist";
+  std::vector<int> nuclides = write_hdf5_nucpath(data_id, nucpath);
```
wanted to use relative path (this is why I stored the different group ids.
But for a reason I ignore it does not work...
So I switched back to absolute path....

bam241 commented:
@gonuke I implemented to change we discussed yesterday...

bam241 committed:
adding some docs aboutwrite_hdf5 algorithm

bam241 committed:
update docs

bam241 committed:
fix algorithm typo

bam241 committed:
autoformat

bam241 committed:
pull change form mat_lib PR

bam241 committed:
cleaning

bam241 commented:
@gonuke I cleaned that up based on you comment on the other PR.

Do you have anymore thoughts  ?

bam241 commented on src/material.cpp:
```diff
@@ -1889,3 +2050,25 @@ pyne::Material pyne::Material::operator/ (double y) {
   return pyne::Material(comp, mass / y, density );
 }
+
+bool pyne::detect_nuclidelist(hid_t data_set, std::string& nucpath){
```
Got this one out of the `Material` class but kept it in the file as it is related to `Material` but I wanted to be able to access it outside of the `Material` class

gonuke commented on src/material.h:
```diff
@@ -151,19 +165,53 @@ namespace pyne
     ///            appended to the end of the dataset.
     /// \param chunksize The chunksize for all material data on disk.
-    void write_hdf5(char * filename, char * datapath, char * nucpath, float row=-0.0,
+    void deprecated_write_hdf5(char * filename, char * datapath, char * nucpath, float row=-0.0,
```
Might need to change the comments to clarify why this is deprecated?

bam241 replied:
shall we move that into `private` to avoid external call ?

gonuke replied:
Sure - if it's not part of the public interface anymore

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
-though they contain methods for converting to/from atom fractions as well.
-In many ways they take inspiration from numpy arrays and python dictionaries.  Materials
-have two main attributes which define them.
+Materials are the primary container for radionuclides. They map nuclides to
```
```suggestion
Materials are the primary container for mixtures of radionuclides. They map nuclides to
```
(I know you didn't change this, but its an opportunity to clarify)

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
-In many ways they take inspiration from numpy arrays and python dictionaries.  Materials
-have two main attributes which define them.
+Materials are the primary container for radionuclides. They map nuclides to
+**mass weights**, though they contain methods for converting to/from atom
```
```suggestion
**mass fractions**, though they contain methods for converting to/from atom
```

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
-1. **comp**: a normalized composition mapping from nuclides (zzaaam-ints) to mass-weights (floats).
-2. **mass**: the mass of the material.
+1. **comp**: a normalized composition mapping from nuclides (zzaaam-ints) to
+   mass-weights (floats).  2. **mass**: the mass of the material.
```
```suggestion
   mass-fractions (floats).  
   2. **mass**: the mass of the material.
```
I'm not sure if this formatting will render correctly

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
+collections of materials.  The material class is presented below.  For more
+information please refer to :ref:`usersguide_material`.
+
+When using the `write_hdf5` method, the default structure for the `HDF5` file
```
```suggestion
When using the `write_hdf5` method to write a material in a group named `my_mat`, the default structure for the `HDF5` file
```

bam241 committed:
improve comments

bam241 committed:
restaure previous format

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
+    /material/my_mat/composition
+    /material/my_mat/nucid
+    /material/my_mat/composition_metadata
+
```
How about something like this?
```
/material/
|--- my_mat/
     |--- composition
     |--- nuclidelist
     |--- composition_metadata
```

gonuke commented on docs/pyapi/material.rst:
```diff
@@ -12,15 +12,38 @@ All functionality may be found in the ``material`` package::
+be provided as the datapath by the user). IF the `$datapath` or `/material`
+exist as a dataset in the file, then the old writing method will ne used.
+
+Older structure are still available to written when providing a `nucpath` to the
```
Is this older structure defined somewhere?

gonuke commented on src/material.cpp:
```diff
@@ -99,59 +116,53 @@ void pyne::Material::_load_comp_protocol1(hid_t db, std::string datapath, int ro
-    comp[nuclides[i]] = (double) (*mat_data).comp[i];
-
+  for (int i = 0; i < nuc_size; i++) {
+    if ((double)(*mat_data).comp[i] != 0) {
```
I thought this changed the other way in the other PR?  If this condition is not met, what do we get for `comp[nuclides[i]]`?

bam241 replied:
this is the composition, I can't find it in the original `material.cpp` file.
I think I introduced this in the MaterialLibrary PR. 
Then removed it into this one, so it disappear from the MaterialLibrary PR and you were highlighted from the merge/rebase commit.

I kinda like it, it prevent to have empty nuclide in a composition. 
`comp` is a `map<nuc, double>` (`nuc` is a `int`) containing the material composition.

gonuke replied:
I see...if the value is zero, we never populate that entry of the dictionary 👍

gonuke commented on src/material.cpp:
```diff
@@ -163,43 +174,39 @@ void pyne::Material::_load_comp_protocol1(hid_t db, std::string datapath, int ro
   H5Tclose(attrtype);
-
-  // Close out the HDF5 file
-  H5Fclose(db);
```
why don't we want to close the file/db here?  We open it in this method, right?

bam241 replied:
we don't here. 
This method takea a `hid_t db` as a parameter so we don't want to close here but in the method that calls it.

gonuke replied:
Yes... I misread this above.

gonuke commented on src/material.cpp:
```diff
@@ -286,29 +285,53 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
     nuc_size = nuclides.size();
     nuc_dims[0] = nuc_size;
+    bool missing_nucs = false;
+    for ( pyne::comp_iter i = comp.begin(); i != comp.end(); i++ ) {
```
Can we assume that either `comp` or `nuclides` is sorted?  Might be able to do a comparison that is more efficient and/or simpler code

bam241 replied:
I don't know enough hdf5 reading/writing mechanism to guaranty it.

they are sorted when we issue the write. so if they remain sorted in the hdf5 dataset and when we read them back, then they are sorted

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +485,163 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+
+    // if datapath != /material and exist -> old format
+    if (h5wrap::path_exists(db, datapath) && datapath != "/material") {
```
I think this might need an additional comment.

If the user gives `datapath = "my_mat"` this `path_exists()` will definitely fail, right?  But if the user gives `datapath = "/my_mat"` then it may succeed as an instance of the old format.  Furthermore, the latter example is never valid in our new structure.  I think this leading slash is a very subtle difference that is worth explaining.

bam241 replied:
I forgot to add the `/` into the `datapath` is not already present...


 it should resolve this potential issue

gonuke replied:
I think it’s probably correct, but it took me a long time of thinking carefully about the difference between the cases I mentioned to convince myself of the behavior

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +485,163 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+        H5O_TYPE_DATASET) {  // "/material" is a dataset -> old format use
+                             // deprecated write_hdf5
+      H5Fclose(db);
+      deprecated_write_hdf5(filename, datapath, "/nucid", row, chunksize);
```
Is it worth it to overload `deprecated_write_hdf5` to have a version that takes a `db` and one that takes a `filename`?

It looks awkward to close this `db` and then open it again in the method

bam241 replied:
as I rename it, I can probably change it to directly take a `db` instead of a filename

gonuke replied:
You might need/want a version that takes a filename as well - at least looking at the use cases above

gonuke commented on news/write_hdf5_structure.rst:
```diff
@@ -0,0 +1,19 @@
+    /material/ 
+    /material/my_mat/ 
+    /material/my_mat/composition
+    /material/my_mat/nucid 
```
```suggestion
    /material/my_mat/nuclidelist 
```

gonuke reviewed (changes requested):
Most of this is documentation, either in comments or in code

bam241 committed:
Apply suggestions from code review

adding changes from @gonuke

Co-Authored-By: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
update docs

bam241 committed:
update docs

bam241 committed:
add deprecation explanation

bam241 committed:
add deprecation explanation

bam241 committed:
add a deprecated write using a db instead of a fileanme

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +485,168 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+    if (!ish5) throw h5wrap::FileNotHDF5(filename);
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+
+    // if datapath != /material and exist -> old format
```
```suggestion
    // there are a few things that may be true here:
    // datapath does not exist - there are many possible reasons for this, including
    //         datapath = "non_default_path" without a leading slash (/) --> assume this is a modifier for a new format
    // datapath does exist:
    //         datapath = "/non_default_path" --> old format
    //         datapath = "/material"   --> new format
```

gonuke commented:
Do you know why this is failing test? I think I'm pretty much ready to merge this

bam241 committed:
Update src/material.cpp

Co-Authored-By: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
add auto-addition of / if not present in datapath

bam241 commented:
> Do you know why this is failing test? I think I'm pretty much ready to merge this

it seems to be the issue #1210  with `LANL` website

bam241 commented:
I also added a quick check to verify that the `datapath` starts a `/`. and had it if not.

gonuke commented:
do we have enough tests of the new behavior?  and how we now treat the old behavior?  etc?

bam241 commented:
> do we have enough tests of the new behavior? and how we now treat the old behavior? etc?

probably not...
I added a test to specifically test old format writing, but don't have test to check all the possible cases when attempting to write in a file that is already in the old format.
I'll add this and ping you.

bam241 committed:
some fix when loading material hdf5 old format adding test to ensure this is done properly

bam241 commented:
@gonuke this should be ok now.

I added tests to verify write and read using old format on the same `datapath` (default and non default path) and on 2 different different paths.
 -- and fixed some part of the from_hdf5()... 

The only thing that we don't (and cannot) check is detecting old format on a non-default `datapath` (if the second write is on a different one) as we don't know where to look for...

bam241 commented on src/material.cpp:
```diff
@@ -221,19 +227,41 @@ void pyne::Material::from_hdf5(std::string filename, std::string datapath, int r
+    H5O_info_t object_info2;
+    status2 = H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+    status2 = H5Oget_info_by_name(db, datapath.c_str(), &object_info, H5P_DEFAULT);
+    // Group "/material" does not exist
```
Had to add a test for H5_DATASET for in the custom `datapath` to the one in the default `datapath` (`/material`)

gonuke commented on src/material.cpp:
```diff
@@ -243,8 +243,13 @@ void pyne::Material::from_hdf5(std::string filename, std::string datapath, int r
     H5O_info_t object_info;
     status = H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
     status = H5Oget_info_by_name(db, "/material" , &object_info, H5P_DEFAULT);
+    herr_t status2;
```
Can we add something like this to help document what we are doing?  i.e. I don't actually know that values of `status` or the logic, but this will make it easier.
```suggestion
    bool material_path_exists = (status ==0 && object_info.type != H5O_TYPE_DATASET)
    herr_t status2;
```
We can then use it in the `if` statement in line 251

gonuke replied:
Similar for status2....

bam241 committed:
Update src/material.cpp

Co-Authored-By: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
propagate change to datapath to check if path exist as a non-dataset

bam241 commented on src/material.cpp:
```diff
@@ -221,19 +227,45 @@ void pyne::Material::from_hdf5(std::string filename, std::string datapath, int r
+    bool datapath_notdataset_exists = (status == 0 && object_info.type != H5O_TYPE_DATASET);
+    
+    // Group "/material" does not exist as a non-dataset
+    if (!material_notdataset_exists || !datapath_notdataset_exists) {
```
@gonuke does it make more sense ?
I used your suggestion, renamed the `bool` variable so it contains all the info and added some comments

gonuke replied:
I think the logic is clearer, but there is some lingering confusion (maybe it's just me) because we are mixing a few cases here.

It is old format if either "/material" or `datapath` is a dataset  - this is clear
It is new format if "/material" is a groupset   - this is clear
It is irrelevant if datapath is a non-dataset   - this is implied but not clear

So maybe some new variable names/logic (pseudocode):

```
enum data_layout_types = {unknown, old_format, new_format};

data_layout_types this_data_layout = unknown

// get info for `datapath`
if (status == 0 && object_info.type == DATASET) this_data_layout = old_format

// get info for "/material"
if (status == 0 && object_info.type == DATASET) { 
    this_data_layout = old_format
} else if (status == 0 && object_info.type == GROUP) {
    this_data_layout = new_format
}

```

In particular, the current code will try the old format if both "/material" and `datapath` do NOT exist.  It will probably fail later, but why not fail here if that is true.

gonuke replied:
instead of the `enum` you could have two `bool`: `format_detected` and `new_format`

bam241 committed:
fix typo in comments:

bam241 committed:
changed the order of testing, we need to keep /material for the else if

bam241 committed:
add comments

bam241 committed:
wording

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +499,176 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+  }
+  // Check if /material exist as a Group
+  // Group "/material/datapath" does not exist create it
+  if (!h5wrap::path_exists(db, "/material" + datapath)) {
```
Isn't your default `datapath` still "/material"?  So doesn't this mean we get "/material/material" as the default group?

bam241 replied:
do you think we should use an other default ? like "mat" ?

gonuke replied:
or "mat_name" or something to make sure it's clear, at least to the software developers...

bam241 committed:
remove old comments

bam241 committed:
new layout detection algo

bam241 commented:
@gonuke I updated the layout detection algorithm.

what do you think ?

bam241 committed:
grammar

gonuke commented on src/material.cpp:
```diff
@@ -221,19 +260,39 @@ void pyne::Material::from_hdf5(std::string filename, std::string datapath, int r
-    _load_comp_protocol1(db, datapath, row);
-  else
+  } else if (protocol == 1) {
+    enum prot1_layout {path_donotexists=-1, unknown, old_layout, new_layout};
```
Define this as a private member of the `Material` class and then use them as the return types of the function?

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +521,175 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+
+    // there are a few things that may be true here:
+    // datapath does not exist - there are many possible reasons for this, including
```
Can we use the new method to determine the layout here as well?

gonuke reviewed (changes requested):
This is a lot better... wonder if we can use it more places - see my comments?

bam241 committed:
re-using detect_hdf5_layout in write_hdf5

bam241 committed:
change from datapath to mat_name

bam241 committed:
add some comment in the write hdf5 switch/case

bam241 commented:
I change in most places `datapath` for `mat_name` except in the deprecated_write_hdf5

I also make use of the `detect_hdf5_layout` method in `write_hdf5`

bam241 committed:
english correction

bam241 commented:
might have discover issues, please restrain from merging yet....

bam241 committed:
this time we are able to write multiple mat in the same file

gonuke commented on src/material.h:
```diff
@@ -123,48 +130,123 @@ namespace pyne
+    ///            appended to the end of the dataset.
+    /// \param chunksize The chunksize for all material data on disk.
+    /// New write_hdf5 which fallback on the old one when required
+    void write_hdf5(std::string filename, std::string mat_name="/material", 
```
I think it was this string "/material" that I was recommending changing to "/mat_name", not the variable name `datapath`.

With this default, you get a group "/material/material" by default which can be a little confusing.

bam241 replied:
I changed it in most of the places... missed that one... :)

bam241 replied:
ok I did not change the default values.... but all the variables name.....

gonuke replied:
I actually think the variable names make more sense as `datapath`....  And we still call it `datapath` in the python layer

gonuke commented on src/material.cpp:
```diff
@@ -469,13 +519,173 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+  
+
+  //mat_name is proided with a "/" so need to open with fullpath
+  // Group "/material/mat_name" does not exist create it
```
You still have "/material" as the default value of the variable `mat_name`, so this comment should refer to "/material/material" (or change the default value?)

bam241 commented:
> might have discover issues, please restrain from merging yet....

I fixed it... but I now am scared to merge without have the test working on CI....

bam241 committed:
change default value of mat_name/datapath

bam241 committed:
change default path to /mat_name in python API and in nosetests

bam241 committed:
revert c++ variabale back to datpath

gonuke commented:
I'm ready to go forward with this one we get passing tests...

bam241 committed:
updating mesh for new material layout

bam241 committed:
cleaning

bam241 commented:
I updated `mesh.py` to use the new material layout.

 the only tests failing on my machine are the `endf` ones.

gonuke commented:
I think some of these failures are real....

bam241 commented:
yes there are,

I started looking at them Friday, did had the time to solve them....

bam241 committed:
detecting matpath location and informing the library opening in mesh

bam241 commented:
@gonuke I fixed the tests.

I add to modify a little bit how `mesh.py` was looking for and loading materials.
see change in [[`mesh.py`](https://github.com/pyne/pyne/pull/1207/files#diff-4be2851467674dbfe9961bc16279fd1b)]

gonuke commented:
Thanks @bam241 !! - this was a long and interesting development and I think we're ready to merge it into the mainline.
